cadence/contracts/bridge/FlowEVMBridge.cdc

```diff
@@ -0,0 +1,405 @@
+    /// Enables all bridge contracts to emit Flow to EVM NFT bridging events from this central contract
+    ///
+    access(account) fun emitBridgeNFTToEVMEvent(
+        type: Type,
+        id: UInt64,
+        evmID: UInt256,
+        to: EVM.EVMAddress,
+        evmContractAddress: EVM.EVMAddress,
+        flowNative: Bool
+    ) {
+        emit BridgedNFTToEVM(type: type, id: id, evmID: evmID, to: to, evmContractAddress: evmContractAddress, flowNative: flowNative)
+        FlowEVMBridgeUtils.call(
+            signature: "emitERC721BridgedFromFlow(address,uint256,address)",
+            targetEVMAddress: FlowEVMBridgeUtils.bridgeFactoryEVMAddress,
+            args: [evmContractAddress, evmID, to],
+            gasLimit: 12000000,
+            value: 0.0
+        )
+    }
+
+    /// Enables all bridge contracts to emit Flow to EVM NFT bridging events from this central contract
+    ///
+    access(account) fun emitBridgeNFTFromEVMEvent(
+        type: Type,
+        id: UInt64,
+        evmID: UInt256,
+        caller: EVM.EVMAddress,
+        evmContractAddress: EVM.EVMAddress,
+        flowNative: Bool
+    ) {
+        emit BridgedNFTFromEVM(type: type, id: id, evmID: evmID, caller: caller, evmContractAddress: evmContractAddress, flowNative: flowNative)
+        FlowEVMBridgeUtils.call(
+            signature: "emitERC721BridgedToFlow(address,uint256,address)",
+            targetEVMAddress: FlowEVMBridgeUtils.bridgeFactoryEVMAddress,
+            args: [evmContractAddress, evmID, caller],
+            gasLimit: 12000000,
+            value: 0.0
+        )
+    }
+
+    /// Handles bridging Flow-native NFTs to EVM - locks NFT in designated Flow locker contract & mints in EVM.
+    /// Within scope, locker contract is deployed if needed & call is passed on to said locker contract.
+    ///
+    access(self) fun bridgeFlowNativeNFTToEVM(
+        token: @{NonFungibleToken.NFT},
+        to: EVM.EVMAddress,
+        tollFee: @FlowToken.Vault
+    ) {
+        let lockerContractName: String = FlowEVMBridgeUtils.deriveLockerContractName(fromType: token.getType()) ??
+            panic("Could not derive locker contract name for token type: ".concat(token.getType().identifier))
+        if self.account.contracts.borrow<&IEVMBridgeNFTLocker>(name: lockerContractName) == nil {
+            self.deployLockerContract(forType: token.getType())
+        }
+
+        let lockerContract: &IEVMBridgeNFTLocker = self.account.contracts.borrow<&IEVMBridgeNFTLocker>(name: lockerContractName)
+            ?? panic("Problem locating Locker contract for token type: ".concat(token.getType().identifier))
+        lockerContract.bridgeNFTToEVM(token: <-token, to: to, tollFee: <-tollFee)
+    }
+    /// Handles bridging Flow-native NFTs from EVM - unlocks NFT from designated Flow locker contract & burns in EVM
+    /// Within scope, locker contract is deployed if needed & passing on call to said contract
+    ///
+    access(self) fun bridgeFlowNativeNFTFromEVM(
```

Comment:
Not sure I understand the question. The bridge abstracts away distinction between "flow-native" and "evm-native" NFTs with a single `bridgeNFTFromEVM` method and conditions calls to `bridgeFlowNativeNFTFromEVM` or `bridgeEVMNativeNFTFromEVM`. The distinction is important as it directs downstream handling of the request boiling down essentiall to - are we burning or transferring in EVM and unlocking or minting in Flow to complete the call?

In #11 I cleaned this path up a bit, but the user-facing interface of `bridgeNFTFromEVM` remains the primary entry point for requests bridging NFTs to Flow from EVM. Ultimately, I wanted to abstract away the need for apps/users to have foreknowledge of where the NFT is originally defined and simply define through the interface the asset and direct they want to bridge to/from.

Does that clear things up or did I miss something?

Reply:
Yes thank you for the explanation. 
I was pointing out the fact that both in/out to/from the bridge are handled on the Cadence side while other bridges provide onramps/offramps on both networks (depending on where the source network is). For example, in an Ethereum Mainnet <-> Flow bridge (with a Flow native asset) you need a Cadence transaction to lock the NFT on Flow side. Then the bridge operator mints your NFT on Ethereum side (with a second transaction on target network). When you want to go back to Flow from Ethereum, you would run a Solidity transaction to transfer the asset to the bridge (and again depend on the bridge admin to unlock your asset on Flow side). It seems like with our approach both operations are handled on Flow side and its okay since EVM environment is inside Cadence environment. I was just wondering the inability to onramp/offramp from Solidity would hurt us in the future or not. I think this is a minor point and the design here makes sense for our problem (we aren't connecting two wholly separated environments).

Reply:
Ah I see what you mean. Yeah, the target for at least the first introduction of the bridge enables moving assets via COA. In the future we might include coverage for bridging via EOAs and to your point would need entry points on either side of the bridge